### PR TITLE
added deck and cards fetching

### PR DESCRIPTION
**Pull request by derricklee43**
Summary of changes:

- Added deck and cards fetching
- Broke page navigation and fetch calls into parent component: `EditDeckPage`
- created a `setDeck` method on `useDeckEditor` aloowing us reset the component even when not unmounted
- Update routes
  - `\edit-deck\:deck-id` a route to an existing deck
  - `\create-deck` a route to a new deck
  - `EditDeckPage` handles both routes and fetches deck if route is `edit-deck`
  - moved all paths to an enum for an easier reference between pages

I removed all navigation and fetching out of the `DeckEditor` to keep the complexity of that component down

https://user-images.githubusercontent.com/39753553/163681192-604bd689-7110-4994-973d-ec79582d3232.mov

### CODE REVIEW

File: src/hooks/api/use-cards-client.ts [hunk 1]
```diff
@@ -46,8 +51,26 @@ export function useCardsClient() {
...
+        text: obj['ftext'],
+        audio: obj['faud'],
+      };
+      card.back = {
+        language: obj['blang'],
+        text: obj['btext'],
+        audio: obj['baud'],
+      };
+      return card;
+    });
```
Comment (derricklee43) [inline src/hooks/api/use-cards-client.ts:71]: You are definitely right about needing a method to convert to JSON. If we convert the Deck to a class and add a static method, we wouldn't have to change the syntax of creating and copying decks, I believe.

File: src/pages/edit-deck-page/edit-deck-page.tsx [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  }, [location]);
+
+  if (deck === undefined) {
+    return <LoadingPage label="loading deck..." />;
+  }
+
+  return (
+    <Fade className="edit-deck-page">
+      <DeckEditor
+        isNewDeck={false}
```
Comment (jiejasonliu) [inline src/pages/edit-deck-page/edit-deck-page.tsx:38]: we'd want to pass `isNewDeck` here right?

File: src/pages/edit-deck-page/edit-deck-page.tsx [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        onCreateDeckClick={handleCreateDeckClick}
+        onDeleteDeckClick={handleDeleteDeckClick}
+        onGoBackClick={navigateBackToDecks}
+      />
+    </Fade>
+  );
+
+  async function fetchDeckAndRefresh() {
+    if (deckId === undefined) {
+      throw new Error('deckId cannot be undefined'); // Todo: maybe route to a 404 page??
```
Comment (jiejasonliu) [inline src/pages/edit-deck-page/edit-deck-page.tsx:49]: definitely something we should do, any thoughts on what our 404 should look like?

File: src/routes.tsx [hunk 4]
```diff
@@ -1,16 +1,26 @@
 import React from 'react';
 import { Routes, Route } from 'react-router-dom';
-import { DeckEditorPage } from './pages/deck-editor/deck-editor';
+import { EditDeckPage } from './pages/edit-deck-page/edit-deck-page';
 import { FlashcardDecksPage } from './pages/decks/flashcard-decks';
 
+export const enum paths {
```
Comment (jiejasonliu) [inline src/routes.tsx:6]: good improvement, i like this 👍

**Review by jiejasonliu — APPROVED** [review]

File: src/hooks/api/use-cards-client.ts (hunk 1 above)
Comment (jiejasonliu) [reply src/hooks/api/use-cards-client.ts:71]: yeah that would definitely make things a lot cleaner

**Comment by jiejasonliu** [comment]
i've also noticed it gets quite confusing when there's a lot of cards and `+ new card` is clicked
maybe we want to [smooth] scroll to the top of the page when it's clicked?

File: src/pages/edit-deck-page/edit-deck-page.tsx (hunk 2 above)
Comment (derricklee43) [reply src/pages/edit-deck-page/edit-deck-page.tsx:38]: yep, thanks for catching that😅

**Comment by derricklee43** [comment]
Yeah, that's I good idea. I'll add some smooth scrolling in another PR

**Pull request merged**